caps/search_funcs.py

```diff
@@ -0,0 +1,171 @@
+    """
+    for p in list_of_phrases:
+        if phrase in p:
+            return True
+    return False
+
+
+def condense_highlights(
+    result: SearchQuerySet, allowed_keywords: list[KeyPhraseSearch] = []
+) -> tuple[SearchQuerySet, list[str]]:
+    """
+    solr's highlighter highlights individual words, even when it's an exact phrase
+    here we're just merging adjacent highlights together
+    """
+    new_result = []
+    extracted_highlights = []
+    for r in result:
+        new_highlights = []
+        for s in r.highlighted.get("text", []):
+            s = s.replace("</mark> <mark>", " ")
+            new_highlights.append(s)
+            # get the text that was highlighted
+            extracted_highlights.extend(
+                [x.lower() for x in re.findall(r"<mark>(.*?)</mark>", s)]
+            )
+        r.highlighted["text"] = new_highlights
+        new_result.append(r)
+    result = new_result
+    matched_keywords = [
+        x
+        for x in allowed_keywords
+        if phrase_is_somewhere_in_list(x.keyphrase, extracted_highlights)
+    ]
+    # sort by cosine similarity
+    matched_keywords = sorted(
+        matched_keywords, key=lambda x: x.cosine_similarity, reverse=True
+    )
+    extracted_highlights = [x.keyphrase for x in matched_keywords]
+
+    return result, extracted_highlights
+
+
+def get_semantic_query(
+    q: str,
+    threshold: float = settings.RELATED_SEARCH_THRESHOLD,
+) -> tuple[Optional[SearchQuerySet], list[KeyPhraseSearch]]:
+    """
+    Run a semantic search query against the contents of the plan documents
+    """
+
+    # get all the related terms
+    all_related = list(KeyPhrase.objects.all())
+    all_related = [x for x in all_related if len(x.keyphrase) > 4]
+    all_related_lookup = {x.id: x for x in all_related}
+    triggered_keywords = []
+    for related in all_related:
+        if related.keyphrase.lower() in q.lower():
+            triggered_keywords.append(related)
+
+    if not triggered_keywords:
+        return None, []
+
+    # for each term, we want to get all the related terms (same set_id)
+    # and then construct a complex query where the original term is replaced by the new term
+    # e.g. if the original query is "housing" and the related term is "housing association"
+    # we want to search for "housing association" OR "housing"
+    all_alternative_terms = KeyPhrasePairWise.objects.filter(
+        word_a_id__in=[x.id for x in triggered_keywords],
+        cosine_similarity__gte=threshold,
+    )
+
+    # group by word_a_id into a dict
+    grouped_related_terms = {
+        k: list(v) for k, v in groupby(all_alternative_terms, lambda x: x.word_a_id)
+    }
+
+    query_parts = []
+    all_related_terms = []
+
+    def get_word_with_cosine(x: KeyPhrasePairWise) -> Optional[KeyPhrase]:
+        """
+        Get back to the phrase object we already have, but keep the cosine
+        from the comparison
+        """
```

Comment:
This is solving a problem where we want the Keyphrase object, but the similarity is stored on the pairwise comparison.

Later on, we want to be able to sort by the similarity when showing which related terms were used - so we can highlight the less expected terms that have been used (e.g. conceptually different, rather than 'light typo').